contrib/annoy-cmake/CMakeLists.txt

```diff
@@ -1,6 +1,6 @@
 option(ENABLE_ANNOY "Enable Annoy index support" ${ENABLE_LIBRARIES})
 
-if ((NOT ENABLE_ANNOY) OR (SANITIZE STREQUAL "undefined"))
+if ((NOT ENABLE_ANNOY) OR (SANITIZE STREQUAL "undefined") OR (SANITIZE STREQUAL "memory"))
```

Comment:
We should fix bugs that have been found by sanitizers.

Reply:
The issue is with the 3rd party "annoylib". We looked into the problems quite extensively in the [original PR](https://github.com/ClickHouse/ClickHouse/pull/37215) and concluded that the failures are false positives. Otherwise, I'd totally agree with you.

Reply:
It could be a false positive. I checked the trace stack and there the error occurs https://github.com/ClickHouse/annoy/blob/9d8a603a4cd252448589e84c9846f94368d5a289/src/annoylib.h#L427 and https://github.com/ClickHouse/annoy/blob/9d8a603a4cd252448589e84c9846f94368d5a289/src/annoylib.h#L398. Both cases were caused by working with node, for which different tricks are used for storage optimality. The memory is initialized with zeros immediately after allocation (https://github.com/ClickHouse/annoy/blob/9d8a603a4cd252448589e84c9846f94368d5a289/src/annoylib.h#L1212), so there is no uninitialized memory there, only zeroing and inserting. Similar errors appeared in UB sanitizer, but that's the way the algorithm is built

Reply:
[Here](https://github.com/ClickHouse/annoy/blob/9d8a603a4cd252448589e84c9846f94368d5a289/src/annoylib.h#L442-L463) the main idea of nodes' storage.

Reply:
Yes, 3rd party libraries usually contain bugs which are found when someone tries to add a library to ClickHouse. And we usually fix such bugs before adding a 3rd party library to our contribs.

> the failures are false positives

Then we should suppress specific false-positive failures (see `__msan_unpoison`, `NO_SANITIZE_UNDEFINED`) instead of disabling sanitizers for the whole library.

Reply:
We tried these fixes (check the last 5-7 comments, which are marked outdated). They lead to errors in std library

Reply:
... because at least `NO_SANITIZE` is not propagated from the annotated function to called functions :-( Didn't try `__msan_unpoison` though. But I would rate littering a 3rd party lib with *san annotations and disabling sanitizers in general equally unbeautiful (here, the "3rd party lib" is a single header).

Reply:
> It could be a false positive. I checked the trace stack and there the error occurs ... 

Seems like you read the report incorrectly. Initially uninitialized value was created on column reallocation in `IColumn::cut` (`Uninitialized value was created by a heap allocation`). Then uninitialized value was [copied](https://github.com/ClickHouse/annoy/blob/9d8a603a4cd252448589e84c9846f94368d5a289/src/annoylib.h#L903-L923) from the column to a "node" in `MergeTreeIndexAggregatorAnnoy::update` (I suppose it actually was out-of-bounds read). Then "node" was copied in `copy_node` (and uninitialized value stored in the "node" was probably copied too). And finally `two_means` read this uninitialized value and tried to actually use that value in branch condition. Using of uninitialized value in branch condition leads to unpredictable (and probably wrong) result, so at this point MSan reports an issue. It does not look like false-positive at all.